### PR TITLE
[feature] add TeaCache support for OmniGen2

### LINKED ISSUE #2371: [RFC]: Support Multi-branch CFG in TeaCache Hook
### Motivation.

The current TeaCache hook's CFG branch management uses hardcoded % 2 logic, only supporting standard 2-branch CFG (positive/negative). Branch identification relies on two fixed strings: "positive" / "negative".

vllm-omni\vllm_omni\diffusion\cache\teacache\hook.py
```
        # ============================================================================
        # GENERIC CACHING LOGIC (works for all models)
        # ============================================================================
        # Set context based on CFG branch for separate state tracking
        # With CFG-parallel, each rank processes only one branch:
        #   - cfg_rank 0: positive branch
        #   - cfg_rank > 0: negative branch
        # Without CFG-parallel, branches alternate within a single rank
        if getattr(module, "do_true_cfg", False):
            cfg_parallel_size = get_classifier_free_guidance_world_size()
            if cfg_parallel_size > 1:
                cfg_rank = get_classifier_free_guidance_rank()
                cache_branch = "negative" if cfg_rank > 0 else "positive"
            else:
                # No CFG-parallel: use forward counter to alternate branches
                cache_branch = "negative" if self._forward_cnt % 2 == 1 else "positive"
        else:
            cache_branch = "positive"
```
However,OmniGen2 uses 3-branch CFG:Each timestep calls the transformer 3 times, but the hook uses % 2 to alternate branch assignment:
timestep 0:
  call 0 (positive)      → 0%2=0 → "positive"
  call 1 (negative+ref)  → 1%2=1 → "negative"
  call 2 (negative)      → 2%2=0 → "positive"  ← WRONG, should be a third branch

### Proposed Change.

Replace the hardcoded 2-branch logic with a generic N-branch approach:

Add ```cfg_num_branches``` attribute, set by the pipeline, default value 2
Sequential mode: replace ```% 2``` with ```% cfg_num_branches```
Parallel mode: use cfg_rank directly as branch index, replacing ```"positive" / "negative"```
Unify branch names to``` f"branch_{idx}"```

```
# Before
if cfg_parallel_size > 1:
    cfg_rank = get_classifier_free_guidance_rank()
    cache_branch = "negative" if cfg_rank > 0 else "positive"
else:
    cache_branch = "negative" if self._forward_cnt % 2 == 1 else "positive"

# After
if cfg_parallel_size > 1:
    cfg_rank = get_classifier_free_guidance_rank()
    cache_branch = f"branch_{cfg_rank}"
else:
    cfg_num_branches = getattr(module, "cfg_num_branches", 2)
    branch_idx = self._forward_cnt % cfg_num_branches
    cache_branch = f"branch_{branch_idx}"
```
### Backward Compatibility
Existing 2-branch CFG models (Qwen-Image, etc.) do not set cfg_num_branches. getattr falls back to default value 2, behavior is unchanged.

### Feedback Period.

_No response_

### CC List.

@princepride @wtomin 

### Any Other Things.

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by zzhuoxin1508**
## Purpose
Add TeaCache support for OmniGen2: register an extractor for `OmniGen2Transformer2DModel` and add its polynomial coefficients, estimated with the official recipe (72 parti-prompts x 50 steps, 3528 rel-L1 pairs, corr=0.94) rather than reusing another model's values.

OmniGen2's image-editing path uses 3-branch CFG (positive + ref_neg + uncond), which relies on the per-branch cache state introduced in #5287 (fixes #2371).
### baseline
python examples/offline_inference/text_to_image/text_to_image.py \
  --model OmniGen2/OmniGen2 \
  --prompt "A curious raccoon wearing a tiny wizard hat, sitting on a stack of ancient books, studio lighting" \
  --height 1024 --width 1024 --num-inference-steps 50 --seed 42 \
  --output omnigen2_baseline.png
### image-edit（3 branchCFG）+ TeaCache + CFG-parallel
python examples/offline_inference/image_to_image/image_edit.py \
  --model OmniGen2/OmniGen2 \
  --image omnigen2_baseline.png \
  --prompt "Change the wizard hat to a red beret" \
  --num-inference-steps 50 --seed 42 \
  --guidance-scale 4.0 --guidance-scale-2 2.0 \
  --cache-backend tea_cache --cfg-parallel-size 3 \
  --output omnigen2_edit_teacache_cfgp3.png

## Test Result
**Text-to-image (2 CFG branches)**:

| config | gen time | speedup | PSNR vs baseline |
|---|---|---|---|
| baseline (no cache) | 31.1s | — | — |
| tea_cache, rel_l1_thresh=0.1 | 20.8s | 1.50x | 33.8 |
| tea_cache, rel_l1_thresh=0.1 + cfg_parallel_size=2 | 10.6s | 2.93x | 33.8 |

**Image editing (3 CFG branches: positive + ref_neg + uncond, image_guidance=2.0)**:

| config | gen time | speedup | PSNR vs edit baseline |
|---|---|---|---|
| edit baseline (no cache) | 89.8s | — | — |
| edit + tea_cache @0.1 | 51.1s | 1.76x | 45.3 |
| edit + tea_cache @0.1 + cfg_parallel_size=2 | 31.0s | 2.90x | 45.2 |
| edit + tea_cache @0.1 + cfg_parallel_size=3 | 24.2s | 3.71x | 45.3 |

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by wtomin** [comment]
A recent PR changed the diffusion features docs strucure. Pls PTAL #1928.

**Comment by zzhuoxin1508** [comment]
> A recent PR changed the diffusion features docs strucure. Pls PTAL #1928.

Noted, thanks!

File: vllm_omni/diffusion/cache/teacache/extractors.py [hunk 1]
```diff
@@ -578,6 +578,183 @@ def postprocess(h):
...
+    """
+    from diffusers.models.modeling_outputs import Transformer2DModelOutput
+    from einops import rearrange
+
+    if not hasattr(module, "layers") or len(module.layers) == 0:
+        raise ValueError("Module must have main transformer layers")
+
+    # ============================================================================
+    # PREPROCESSING (OmniGen2-specific)
+    # ============================================================================
```
Comment (lishunyang12) [inline vllm_omni/diffusion/cache/teacache/extractors.py:617]: Duplicated comment block — the `# PREPROCESSING` header + separator appears twice.

```suggestion
    # PREPROCESSING (OmniGen2-specific)
    # ============================================================================
```

File: vllm_omni/diffusion/cache/teacache/extractors.py [hunk 2]
```diff
@@ -578,6 +578,183 @@ def postprocess(h):
...
+            "attention_mask": attention_mask,
+            "rotary_emb": rotary_emb,
+            "seq_lengths": seq_lengths,
+            "l_effective_img_len": l_effective_img_len,
+            "img_sizes": img_sizes,
+            "is_hidden_states_tensor": is_hidden_states_tensor,
+        },
+    )
+
+
```
Comment (lishunyang12) [inline vllm_omni/diffusion/cache/teacache/extractors.py:746]: None of these `extra_states` keys are read by the hook or anywhere else — `run_transformer_blocks` and `postprocess` already capture everything they need via closure. This just allocates a dict that sits unused. Remove it or set `extra_states=None`.

File: vllm_omni/diffusion/cache/teacache/config.py [hunk 3]
```diff
@@ -37,6 +37,15 @@
     # Bagel transformer coefficients
     # Using Qwen's coefficients as reasonable default given shared architecture
     "Bagel": [1.33313129e06, -1.68644226e05, 7.95050740e03, -1.63747873e02, 1.26352397e00],
```
Comment (lishunyang12) [inline vllm_omni/diffusion/cache/teacache/config.py:39]: The test plan says results will be provided "as soon as possible" — are these coefficients validated at all yet? Shipping untuned polynomial coefficients could silently degrade output quality. Would be good to at least confirm basic generation quality before merging.

**Review by lishunyang12 — COMMENTED** [review]
left a couple comments

**Comment by zzhuoxin1508** [comment]
Sorry for the delay.Right now TeaCache only supports 2-branch CFG ,but omnigen2 is  3-branch  . I've filed this as issue#2371. So these coefficients only cover the 2-branch path; fuller validation depends on #2371. @lishunyang12

File: vllm_omni/diffusion/cache/teacache/extractors.py (hunk 1 above)
Comment (zzhuoxin1508) [reply vllm_omni/diffusion/cache/teacache/extractors.py:617]: fixed

File: vllm_omni/diffusion/cache/teacache/extractors.py (hunk 2 above)
Comment (zzhuoxin1508) [reply vllm_omni/diffusion/cache/teacache/extractors.py:746]: fixed

**Comment by wtomin** [comment]
Is there any follow-up PRs related to #2371?

**Comment by wtomin** [comment]
@zzhuoxin1508 I think this PR is stale for about three months. If no further updates, I will close this PR in a few days.

**Comment by zzhuoxin1508** [comment]
> @zzhuoxin1508 I think this PR is stale for about three months. If no further updates, I will close this PR in a few days.

@wtomin Sorry for the long silence, and thanks for the ping. Yes, #2371 is doable — it's the blocker for this PR, since OmniGen2 uses 3-branch CFG while the TeaCache hook currently only handles 2 branches. I'll get both #2371 and this PR done within this week. Please keep it open, thanks!

File: vllm_omni/diffusion/cache/teacache/config.py (hunk 3 above)
Comment (zzhuoxin1508) [reply vllm_omni/diffusion/cache/teacache/config.py:39]: fixed

**Review by RuixiangMa — COMMENTED** [review]
@zzhuoxin1508 please add ut for the pr

File: vllm_omni/diffusion/cache/teacache/extractors.py [hunk 4]
```diff
@@ -577,6 +577,173 @@ def postprocess(h):
     )
 
 
+def extract_omnigen2_context(
```
Comment (RuixiangMa) [inline vllm_omni/diffusion/cache/teacache/extractors.py:580]: TeaCache hook replaces forward entirely , so the extractor must replicate all logic from the original forward. The original calls sp_input_boundary / sp_output_boundary around the layer loop — these are no-op modules that serve as SP hook attachment points (_sp_plan). The extractor's run_transformer_blocks only has the layer loop, skipping both. Suggest to  add module.sp_input_boundary / module.sp_output_boundary to run_transformer_blocks.

Comment (zzhuoxin1508) [reply vllm_omni/diffusion/cache/teacache/extractors.py:580]: Fixed — run_transformer_blocks now goes through both SP boundaries, with a UT covering this.

**Comment by zzhuoxin1508** [comment]
@RuixiangMa Added TestOmniGen2Extractor with an extractor-vs-native forward equivalence test.
